File: packages/pdf-annotator/android/src/main/java/io/capawesome/capacitorjs/plugins/pdfannotator/PdfAnnotatorPlugin.java
package io.capawesome.capacitorjs.plugins.pdfannotator;

import android.content.Intent;
import androidx.activity.result.ActivityResult;
import androidx.annotation.NonNull;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.pdfannotator.classes.CustomException;
import io.capawesome.capacitorjs.plugins.pdfannotator.classes.options.OpenOptions;
import io.capawesome.capacitorjs.plugins.pdfannotator.classes.results.IsAvailableResult;
import io.capawesome.capacitorjs.plugins.pdfannotator.classes.results.OpenResult;
import io.capawesome.capacitorjs.plugins.pdfannotator.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.pdfannotator.interfaces.Result;

@CapacitorPlugin(name = "PdfAnnotator")
public class PdfAnnotatorPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "PdfAnnotatorPlugin";

    private PdfAnnotator implementation;

    @PluginMethod
    public void isAvailable(PluginCall call) {
        try {
            NonEmptyResultCallback<IsAvailableResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull IsAvailableResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.isAvailable(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @Override
    public void load() {
        super.load();
        this.implementation = new PdfAnnotator(this);
    }

    @PluginMethod
    public void open(PluginCall call) {
        try {
            OpenOptions options = new OpenOptions(call);
            Intent intent = implementation.createOpenIntent(options);
            startActivityForResult(call, intent, "handleOpenResult");
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @ActivityCallback
    private void handleOpenResult(PluginCall call, ActivityResult result) {
        if (call == null) {
            return;
        }
        try {
            NonEmptyResultCallback<OpenResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull OpenResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.handleOpenResult(result, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call, @NonNull Result result) {
        call.resolve(result.toJSObject());
    }
}

File: packages/pdf-annotator/ios/Plugin/PdfAnnotatorPlugin.swift
import Foundation
import Capacitor

@objc(PdfAnnotatorPlugin)
public class PdfAnnotatorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PdfAnnotatorPlugin"
    public let jsName = "PdfAnnotator"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise)
    ]

    public static let tag = "PdfAnnotatorPlugin"

    private var implementation: PdfAnnotator?

    @objc func isAvailable(_ call: CAPPluginCall) {
        implementation?.isAvailable(completion: { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        })
    }

    override public func load() {
        self.implementation = PdfAnnotator(plugin: self)
    }

    @objc func open(_ call: CAPPluginCall) {
        do {
            let options = try OpenOptions(call)

            implementation?.open(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call, result)
            })
        } catch {
            rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", PdfAnnotatorPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
